src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Download/PackageDownloadRunner.cs

```
@@ -238,53 +238,71 @@ private static bool TryGetRepositoriesForPackage(
-                foreach (var mappedName in mappedNames)
-                {
-                    SourceRepository? repo = null;
-                    for (int i = 0; i < allRepos.Count; i++)
-                    {
-                        if (string.Equals(allRepos[i].PackageSource.Name, mappedName, StringComparison.OrdinalIgnoreCase))
-                        {
-                            repo = allRepos[i];
-                            break;
-                        }
-                    }
+                resultRepositories = GetMappedRepositories(mappedNames, allRepos, packageId, logger);
 
-                    if (repo != null)
-                    {
-                        mappedRepos.Add(repo);
-                    }
-                    else
-                    {
-                        logger.LogVerbose(
-                            string.Format(
-                                CultureInfo.CurrentCulture,
-                                Strings.PackageDownloadCommand_PackageSourceMapping_NoSuchSource,
-                                mappedName,
-                                packageId));
-                    }
-                }
-
-                if (DetectAndReportInsecureSources(args.AllowInsecureConnections, mappedRepos.Select(repo => repo.PackageSource), logger))
+                if (DetectAndReportInsecureSources(args.AllowInsecureConnections, resultRepositories.Select(repo => repo.PackageSource), logger))
                 {
                     repositories = [];
                     return false;
                 }
-
-                repositories = mappedRepos;
-                return true;
             }
             else
             {
-                // No mapping for this package: fall back to all sources
-                repositories = allRepos;
-                return true;
+                // No mapping for this package
+                resultRepositories = allRepos;
```

Shouldn't we call DetectAndReportInsecureSources for allRepos here?

---

Great catch! That was a bug. Updated and tested

---

Do we need to add a test to catch an insecure source both for mapped sources and for regular sources, so that this would have been caught through UTs?

---

Update: I have removed this scenario. The bug is that if source mapping is enabled, and a mapped source was found for a package, we fail. That means we don't need to do the insecure source check in this scenario anymore. More https://github.com/NuGet/NuGet.Client/pull/6953#discussion_r2544001974


regarding testing for insecure source both for mapped sources and for regular sources,
we have the following: https://github.com/NuGet/NuGet.Client/blob/7bd2e43dc81c5a2df40e7137574d0f143c03f4fc/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Package/Download/PackageDownloadRunnerTests.cs#L752 https://github.com/NuGet/NuGet.Client/blob/7bd2e43dc81c5a2df40e7137574d0f143c03f4fc/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Package/Download/PackageDownloadRunnerTests.cs#L374
https://github.com/NuGet/NuGet.Client/blob/7bd2e43dc81c5a2df40e7137574d0f143c03f4fc/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Package/Download/PackageDownloadRunnerTests.cs#L591
https://github.com/NuGet/NuGet.Client/blob/7bd2e43dc81c5a2df40e7137574d0f143c03f4fc/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/Package/Download/PackageDownloadRunnerTests.cs#L644

---

I see that comment as unrelated to the check for insecure sources.

@nkolev92  - thoughts?

---

Thanks for the follow up! Just to clarify the context of my earlier comment: the **update** I linked was specifically addressing the scenario from the original discussion about whether we should run the insecure-source check over all repositories.

Previously, when PSM was enabled and a package had no matching mappings, the old behavior was to fall back to using all sources. Your original question about “Shouldn't we call DetectAndReportInsecureSources for allRepos here” applied to that fallback path.

With the new implementation, we no longer fall back to all sources when a package has no mapping. Instead, if PSM is enabled and a mapped doesn't exist, we fail early. Because that fallback no longer happens, the “check allRepos for insecure sources” scenario doesn’t apply anymore

In the second part of my comment, I was addressing your other question regarding insecure-source testing.